### DISCUSSION TITLE
Is Lua the right language?

### DISCUSSION THREAD

**Discussion by Fail-Safe** [sqm-autorate/sqm-autorate#88] [General]
Obviously we are making some neat things happen with Lua, but given the large spans of time between various Lua package updates, Lanes having to be recompiled for LuaJIT, etc., I can't help but wonder if Lua is the right thing in 2022.

Is there any consideration around what, if any, other language(s) might be more ideal for future growth and development?

**Comment by Fail-Safe** [comment]
What about Go? Would it make sense to develop in Go and cross-compile for multiple platforms?

**Reply by Lochnair** [reply]
I did briefly consider Go for this, but to my knowledge the binary files take quite a bit of space. At least they used to.
My goal has been to have it run on any _supported_ OpenWrt device, that means 8M flash devices and up these days. Which is not a lot.

**Reply by Fail-Safe** [reply]
Understood and I definitely agree with the most minimal, yet functional approach.

I'm going to explore this a bit more just for the fun of it. I've dabbled in learning Go a few times in the past and have been looking for a "real" reason to just dive in and get my hands dirty with it.

Just for grins, any clue what an x86 binary size comes to for the very most basic "hello world" app in Go?

**Reply by Lochnair** [reply]
I have a clue now :P

```go
package main

import (
	"fmt"
)


func main() {
	fmt.Println("hello world")
}
```

Compiled with Go 1.17.6, that turns into a 1.7M large binary, note that this includes debugging symbols, even so.
Stripping debugging information from the binary gets us down to 1.2M. That's still quite a lot.

**Reply by Fail-Safe** [reply]
Yeah, that's pretty big 👎  Well maybe that's enough to know for now and we'll put this back on the shelf.

Are there any other languages, compiled or otherwise, that would make sense to explore?

**Reply by jeverley** [reply]
I pondered the same myself, I know that user Grommish is also working on rust support within OpenWrt, but still looks like a work in progress - a possibility depending on its maturity once we've firmed up algorithmic behaviours etc.
[https://forum.openwrt.org/t/rust-lang-rustc-cargo-for-openwrt-testing-needed/71157/74](url)
[https://github.com/openwrt/packages/pull/13916](url)

**Comment by dlakelan** [comment]
There is tinygo https://tinygo.org/ (I'm planning to try to use it for some RPi Pico type stuff)

but it's really targeting microcontrollers. I don't know if it can be made to target Linux on ARM or Linux on MIPS. Someone want to try?

Possibly this could be done by targetting WASM and then running it on wasm3 interpreter on the router?

**Reply by dlakelan** [reply]
https://github.com/tinygo-org/tinygo/issues/1075

indicates that people are able to get tinygo to run on MIPS so, I think this is a very doable thing. However, it makes the whole thing a lot less "tinkerable" for the user compared to an interpreted language solution.

**Reply by Fail-Safe** [reply]
Not to lose sight of the "less tinkerable" theme, but FYI...

Simple "hello world":
```go
package main

import "fmt"

func main() {
	fmt.Println("Hello world!")
}
```

Compiled with Go 1.17.5 on MacOS:
```shell
❯ ls -lart
total 3664
drwxr-xr-x  31 mark  staff      992 Jan 11 19:52 ..
-rw-r--r--   1 mark  staff       73 Jan 11 19:53 hello.go
-rwxr-xr-x   1 mark  staff  1869504 Jan 11 19:53 hello
drwxr-xr-x   4 mark  staff      128 Jan 11 19:53 .
```

Compiled with TinyGo 0.21.0 on MacOS:
```shell
❯ ls -lart
total 192
drwxr-xr-x  31 mark  staff    992 Jan 11 19:52 ..
-rw-r--r--   1 mark  staff     73 Jan 11 19:53 hello.go
-rwxr-xr-x   1 mark  staff  94000 Jan 11 19:53 hello
drwxr-xr-x   4 mark  staff    128 Jan 11 19:53 .
```

Works!
```shell
❯ ./hello
Hello world!
```

That's about a 95% reduction in size if I'm doing the math right. That at least gets us into a theoretical ballpark for 8MB device potential.

**Reply by Fail-Safe** [reply]
After stripping the binary, it gets even smaller:

```shell
❯ strip -S hello
❯ ls -lart
total 176
drwxr-xr-x  31 mark  staff    992 Jan 11 19:52 ..
-rw-r--r--   1 mark  staff     73 Jan 11 19:53 hello.go
-rwxr-xr-x   1 mark  staff  82816 Jan 11 20:14 hello
drwxr-xr-x   4 mark  staff    128 Jan 11 20:14 .
```

**Reply by dlakelan** [reply]
Love it! Not sure how well it would work for MIPS/ARM/PPC etc but my suspicion is it'll work with appropriate tinkering.

**Reply by Fail-Safe** [reply]
I'll see if I can cross-compile this "hello world" for ARM64 tomorrow and check the size of it.

**Reply by Fail-Safe** [reply]
I went ahead and tested a cross-compile for ARM64:

```shell
❯ env GOOS=linux GOARCH=arm64 tinygo build -o hello ./hello.go
❯ ls -lart
total 616
drwxr-xr-x  31 mark  staff     992 Jan 11 19:52 ..
-rw-r--r--   1 mark  staff      73 Jan 11 19:53 hello.go
-rwxr-xr-x   1 mark  staff  310016 Jan 11 23:36 hello
drwxr-xr-x   4 mark  staff     128 Jan 11 23:36 .
❯ file hello
hello: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, with debug_info, not stripped
❯ upx --brute hello
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2020
UPX 3.96        Markus Oberhumer, Laszlo Molnar & John Reiser   Jan 23rd 2020

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
    310016 ->    102740   33.14%   linux/arm64   hello

Packed 1 file.
❯ ls -lart
total 216
drwxr-xr-x  31 mark  staff     992 Jan 11 19:52 ..
-rw-r--r--   1 mark  staff      73 Jan 11 19:53 hello.go
-rwxr-xr-x   1 mark  staff  102740 Jan 11 23:36 hello
drwxr-xr-x   4 mark  staff     128 Jan 11 23:37 .
❯ file hello
hello: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, no section header
```

**Reply by Fail-Safe** [reply]
@Lochnair I ran up against this when trying to compile for MIPS last night:
```shell
❯ env GOOS=linux GOARCH=mips tinygo build -o hello ./hello.go
error: open /usr/local/Cellar/tinygo/0.21.0/lib/musl/arch/mips/bits/alltypes.h.in: no such file or directory
```

Looks like the link @dlakelan sent (tinygo-org/tinygo#1075) is the key, but I didn't run too far down that trail last night. I'll try again later today, but if you beat me to it, let me know!

**Reply by Lochnair** [reply]
@Fail-Safe Yeah I read through that issue, didn't look promising at all with the syscall problems he had. So while tinygo looks really neat I think it'll be too much work to get it running on all the architectures we'd need to support.

I did toy with the idea of compiling tinygo to WebAssembly and throwing the wasm3 interpreter in OpenWrt to run it, but I haven't gotten anywhere with it even locally on my x86_64 box.

**Reply by Fail-Safe** [reply]
Hopefully as tinygo continues to evolve and improve, we can revisit it in the future. Sounds like at this point Lua stays the course for us. I will say, this news of LuaJIT is very promising: https://github.com/Fail-Safe/sqm-autorate/discussions/76#discussioncomment-1954727
